dio/lib/src/transformers/utf8_json_transformer.dart

```diff
@@ -0,0 +1,184 @@
+    // and the response is json -> use the fast path decoder
+    if (isJsonContent && customResponseDecoder == null) {
+      return _fastUtf8JsonDecode(responseBody);
+    }
+    final responseBytes = await _consolidateStream(responseBody.stream);
+
+    // a custom response decoder overrides the default behavior
+    final String? decodedResponse;
+
+    if (customResponseDecoder != null) {
+      final decodeResponse = customResponseDecoder(
+        responseBytes,
+        options,
+        responseBody..stream = const Stream.empty(),
+      );
+
+      if (decodeResponse is Future) {
+        decodedResponse = await decodeResponse;
+      } else {
+        decodedResponse = decodeResponse;
+      }
+    } else {
+      decodedResponse = null;
+    }
+
+    if (isJsonContent && decodedResponse != null) {
+      // slow path decoder, since there was a custom decoder specified
+      return jsonDecode(decodedResponse);
+    } else if (decodedResponse != null) {
+      return decodedResponse;
+    } else {
+      // If the response is not JSON and no custom decoder was specified,
+      // assume it is an utf8 string
+      return utf8.decode(
+        responseBytes,
+        allowMalformed: true,
+      );
+    }
+  }
+
+  Future<Object?> _fastUtf8JsonDecode(ResponseBody responseBody) async {
+    final contentLengthHeader =
+        responseBody.headers[Headers.contentLengthHeader];
+
+    final hasContentLengthHeader =
+        contentLengthHeader != null && contentLengthHeader.isNotEmpty;
+
+    // the content length of the response, either from the content-length header
+    // of the response or the length of the eagerly decoded response bytes
+    final int contentLength;
+
+    // the eagerly decoded response bytes
+    // which is set if the content length is not specified
+    // null otherwise (we'll feed the stream directly to the decoder in that case)
+    Uint8List? responseBytes;
+
+    // if the content length is not specified, we need to consolidate the stream
+    // and count the bytes to determine if we should use an isolate
+    // otherwise we use the content length header
+    if (!hasContentLengthHeader) {
+      responseBytes = await _consolidateStream(responseBody.stream);
+      contentLength = responseBytes.length;
+    } else {
+      contentLength = int.parse(contentLengthHeader.first);
+    }
+
+    final shouldUseIsolate = !(contentLengthIsolateThreshold < 0) &&
+        contentLength >= contentLengthIsolateThreshold;
+    if (shouldUseIsolate) {
+      // we can't send the stream to the isolate, so we need to decode the response bytes first
+      return compute(
+        _decodeUtf8ToJson,
+        responseBytes ?? await _consolidateStream(responseBody.stream),
+      );
+    } else {
+      if (!hasContentLengthHeader || contentLength == 0) {
+        responseBytes ??= await _consolidateStream(responseBody.stream);
+        // if the response is empty, return null, since the decoder would throw
```

Comment:
Could you explain more on this case with examples? Also comments should start with capital letter and end with dot too:

```suggestion
        // Returns `null` to let the decoder throws when the response is empty.
```

Reply:
I added a more detailed explanation as comment.
This branch is for backwards compatibility with the other Transformers.

The old behavior was, that if a server responds with an `Content-Type: application/json` header. but the response body was empty, the transformer would return null. 
But the fused decoder would throw an exception in this case.

So, if we don't have a valid Content-Length header, we first check if the body is empty before converting the bytes to json.
If we do have a Content-Length header which is >0, we can feed the response bytes directly into the decoder, which is the most performant path.